lib/src/client.dart

```diff
@@ -2466,50 +2472,58 @@ class Client extends MatrixApi {
-              stateEvent.relationshipType != RelationshipTypes.edit;
-          final editingLastEvent =
-              stateEvent.relationshipEventId == room.lastEvent?.eventId;
-          final consecutiveEdit =
-              room.lastEvent?.relationshipType == RelationshipTypes.edit &&
-                  stateEvent.relationshipEventId ==
-                      room.lastEvent?.relationshipEventId;
-          final importantOrRoomLoaded =
-              eventUpdate.type == EventUpdateType.inviteState ||
-                  !room.partial ||
-                  // make sure we do overwrite events we have already loaded.
-                  room.states[stateEvent.type]
-                          ?.containsKey(stateEvent.stateKey ?? '') ==
-                      true ||
-                  importantStateEvents.contains(stateEvent.type);
-          if ((noMessageOrNoEdit || editingLastEvent || consecutiveEdit) &&
-              importantOrRoomLoaded) {
-            room.setState(stateEvent);
+        room.setState(Event.fromJson(eventUpdate.content, room));
+        break;
+      case EventUpdateType.state:
+      case EventUpdateType.timeline:
+        final event = Event.fromJson(eventUpdate.content, room);
+
+        // Update the room state:
+        if (!room.partial ||
+            // make sure we do overwrite events we have already loaded.
+            room.states[event.type]?.containsKey(event.stateKey ?? '') ==
+                true ||
+            importantStateEvents.contains(event.type)) {
+          room.setState(event);
+        }
+        if (eventUpdate.type != EventUpdateType.timeline) break;
+
+        // If last event is null or not a valid room preview event anyway,
+        // just use this:
+        if (room.lastEvent == null ||
+            !roomPreviewLastEvents.contains(room.lastEvent?.type)) {
+          room.lastEvent = event;
+          break;
+        }
+
+        // Is this event of an important type for the last event?
+        if (!roomPreviewLastEvents.contains(event.type)) break;
+
+        // Is this event redacting the last event?
+        if (event.type == EventTypes.Redaction &&
+            event.redacts == room.lastEvent?.eventId) {
+          room.lastEvent?.setRedactionEvent(event);
+          break;
+        }
+
+        // Is this event an edit of the last event? Otherwise ignore it.
+        if (event.relationshipType == RelationshipTypes.edit) {
+          if (event.relationshipEventId == room.lastEvent?.eventId ||
+              (room.lastEvent?.relationshipType == RelationshipTypes.edit &&
+                  event.relationshipEventId ==
+                      room.lastEvent?.relationshipEventId)) {
+            room.lastEvent = event;
           }
+          break;
         }
+
+        // Ignore reactions for last event.
+        if (event.relationshipType == RelationshipTypes.reaction) break;
+
+        // Event is a valid new lastEvent:
+        room.lastEvent = event;
```

Comment:
So, this leaves me a bit confused:

- You set the last event to event, if it is null (makes sense)
- You set the last event if the last event is not a preview event (makes sense)
- Then you skip everything, that isn't a room preview event (this seems to early to me)
- If the event is a redaction, it redacts the last event (good, but you already skipped redactions, since they are not in the preview events)
- If it is an edit, update last event (good)
- You skip reactions (good, but we never should have any normal reactions there, since their event type was not in the roomPreviewEvents)
- Then you set the last event

I guess most of that is fine, the only thing probably an issue is that redactions would need to be handled before the previewEvents check and that we might have a weird preview in that case?

Reply:
Thanks. You are right. I have fixed it by move the check if type is a room preview event type below the check for the redaction and removed the check for reaction as this is handled by the room preview event type anyway.